functions.php

```diff
@@ -0,0 +1,282 @@
+<?php
+
+function logErrors(string $customMessage, string $errorMessage, int $line): void
+{
+    error_log(
+        date('Y-m-d H:i:s') . " - Line: {$line} = {$customMessage} {$errorMessage} " . PHP_EOL,
+        3,
+        'errors.txt'
+    );
+}
+
+function toJson(array $data): ?string
+{
+    try {
+        $jsonEncoded = json_encode($data, JSON_THROW_ON_ERROR, 512);
+    } catch (JsonException $e) {
+        logErrors('Error encoding json', $e->getMessage(), __LINE__);
+        return null;
+    }
+
+    return $jsonEncoded;
+}
+
+function fromJson(string $data): ?array
+{
+    try {
+        $jsonDecoded = json_decode($data, true, 512, JSON_THROW_ON_ERROR);
+    } catch (JsonException $e) {
+        logErrors('Error decoding json', $e->getMessage(), __LINE__);
+        return null;
+    }
+
+    return $jsonDecoded;
+}
+
+function sendResponse(string $message, int $statusCode): void
+{
+    http_response_code($statusCode);
+
+    echo $message;
+
+    exit;
+}
+
+/**
+ * @return array{endpoint:string,params:array,method:string}
+ */
+function processURL(): array
+{
+    $request = parse_url($_SERVER['REQUEST_URI']);
+
+    $params = [];
+
+    if (isset($request['query'])) {
+        parse_str($request['query'], $params);
+    }
+
+    return [
+        'endpoint' => str_replace('/', '', $request['path']), 
+        'params' => $params,
+        'method' => $_SERVER['REQUEST_METHOD'], 
+    ];
+}
+
+function readProducts(string $path): ?array
+{
+    $productsRead = @file_get_contents($path);
+
+    if ($productsRead === false) {
+        return null;
+    }
+
+    return fromJson($productsRead);
+}
+
+function extractProduct(array $products, string $category, ?int $id = null): ?array
+{
+    if (!isset($products[$category])) {
+        return null;
+    }
+
+    if ($id !== null) {
+        foreach ($products[$category] as $product) {
+            if ($product['id'] == $id) {
+                return $product;
+            }
+        }
+
+        return null;
+    }
+
+    return $products[$category];
```

Comment:
```suggestion
    if ($id === null) {
        return $products[$category];
    }

    foreach ($products[$category] as $product) {
        if ($product['id'] == $id) {
            return $product;
        }
    }

    return null;
```

Reply:
@AndreiOrmanji - Added suggested change to the `extractProduct()` function.